### PR TITLE
[SYSTEMDS-2669] Python Mnist LogReg Tutorial

### PR DESCRIPTION
**Pull request by Baunsgaard**
Python Tutorial for Mnist dataset using LogReg as the algorithm.

This pr also adds:

- One Hot Encoding in python
- an Mnist data-set parser for downloading and parsing Mnist easily
- moving matrix python tests to own folder.

### CODE REVIEW

**Comment by Baunsgaard** [comment]
Also added shape as a variable for the OperatorNode, this is not a finished concept yet since reading files from disk does not contain the shapes, and multi return does not return the shapes correctly either.

Future work would be to improve the design of the multi returns such that one is able to use individual parts of these in subsequent operations.

**Comment by Baunsgaard** [comment]
This should be ready for review now.

**Comment by sebwrede** [comment]
The tests are failing, @Baunsgaard.

File: src/main/python/docs/source/guide/algorithms.rst [hunk 1]
```diff
@@ -26,18 +26,176 @@ Prerequisite:
 
 - :doc:`/getting_started/install`
 
+This example goes through an algorithm from the list of builtin algorithms that can be applied to a dataset.
+For simplicity the dataset used for this is `MNist <http://yann.lecun.com/exdb/mnist/>`_,
+since it is commonly known and explored.
+
+If one want to skip the explanation then the full script is available in to bottom of this page.
```
Comment (sebwrede) [inline src/main/python/docs/source/guide/algorithms.rst:33]: > If one want* to

*wants 

> available (in to)* bottom
*at the

File: src/main/python/docs/source/guide/algorithms.rst [hunk 2]
```diff
@@ -26,18 +26,176 @@ Prerequisite:
...
 
 Step 1: Get Dataset
 -------------------
 
-TODO
+Systemds provide builtin for downloading and setup of the mnist dataset.
```
Comment (sebwrede) [inline src/main/python/docs/source/guide/algorithms.rst:38]: > Systemds provide*
*provides

File: src/main/python/docs/source/guide/algorithms.rst [hunk 3]
```diff
@@ -26,18 +26,176 @@ Prerequisite:
...
+To setup this simply use::
 
-Step 2: Train model
--------------------
+    from systemds.examples.tutorials.mnist import DataManager
+    d = DataManager()
+    X = d.get_train_data()
```
Comment (sebwrede) [inline src/main/python/docs/source/guide/algorithms.rst:43]: I get: 

`>>> X = d.get_train_data()

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.local/lib/python3.8/site-packages/systemds/examples/tutorials/mnist.py", line 56, in get_train_data
    self._get_data(self._train_data_url, self._train_data_loc)
  File "/.local/lib/python3.8/site-packages/systemds/examples/tutorials/mnist.py", line 102, in _get_data
    os.mkdir(folder)
FileNotFoundError: [Errno 2] No such file or directory: 'systemds/examples/tutorials/mnist'
`

File: src/main/python/docs/source/guide/algorithms.rst [hunk 4]
```diff
@@ -26,18 +26,176 @@ Prerequisite:
...
+Here the DataManager contains the code for downloading and setting up numpy arrays containing the data.
+
+Step 2: Reshape & Format
+------------------------
+
+Usually data does not come in formats that perfectly fits the algorithms, to make this tutorial more
+realistic some data preprocessing is required to change the input to fit.
+
+First the Training data, X, has multiple dimensions resulting in a shape (60000, 28, 28).
+these dimensions corresponds to first the number of images 60000, then the number of row pixels, 28,
```
Comment (sebwrede) [inline src/main/python/docs/source/guide/algorithms.rst:55]: *These

Comment (sebwrede) [reply src/main/python/docs/source/guide/algorithms.rst:55]: *correspond

File: src/main/python/docs/source/guide/algorithms.rst [hunk 5]
```diff
@@ -26,18 +26,176 @@ Prerequisite:
...
+and finally the column pixels, 28.
+
+To use this data for Logistic Regression we have to reduce the dimensions.
+The input, X, of this algorithm require the data, to have two dimensions, and the first resemble the
```
Comment (sebwrede) [inline src/main/python/docs/source/guide/algorithms.rst:59]: Check commas

**Review by sebwrede — COMMENTED** [review]
This is only the first part of the review. I cannot import SystemDS correctly in Python when I follow the tutorial. Do you know why, @Baunsgaard? See my comment with the error message. 
Another general comment: write "SystemDS" and not "systemds" unless it is a command where it is important that the characters are not capitalized. You could even write "Apache SystemDS" :smile:

**Comment by Baunsgaard** [comment]
> This is only the first part of the review. I cannot import SystemDS correctly in Python when I follow the tutorial. Do you know why, @Baunsgaard? See my comment with the error message.
> Another general comment: write "SystemDS" and not "systemds" unless it is a command where it is important that the characters are not capitalized. You could even write "Apache SystemDS" smile

You need to install systemds python if you want to run it.
therefore i refer to the install guide that has to be run first.

File: src/main/python/docs/source/guide/algorithms.rst (hunk 3 above)
Comment (Baunsgaard) [reply src/main/python/docs/source/guide/algorithms.rst:43]: Should be fixed, The issue is that the mkdir method does not create nester folders, i replaced it with mkdirs.

File: src/main/python/docs/source/guide/algorithms.rst [hunk 6]
```diff
@@ -26,18 +26,177 @@ Prerequisite:
...
+
+With these steps we are now ready to train a simple logistic model.
+
+Step 3: Training
+----------------
+
+To start with, we setup a SystemDS context::
+
+    from systemds.context import SystemDSContext
+    sds = SystemDSContext()
```
Comment (sebwrede) [inline src/main/python/docs/source/guide/algorithms.rst:85]: `File "/systemds/src/main/python/systemds/context/systemds_context.py", line 96, in __init__
    raise Exception("Exception in startup of GatewayServer: " + stderr)
Exception: Exception in startup of GatewayServer: Error: Could not find or load main class org.apache.sysds.api.PythonDMLScript
`

Comment (sebwrede) [reply src/main/python/docs/source/guide/algorithms.rst:85]: It is required to run `export SYSTEMDS_ROOT=$(pwd)` in SystemDS root before doing this step.

Comment (Baunsgaard) [reply src/main/python/docs/source/guide/algorithms.rst:92]: should be resolved now.

**Comment by Baunsgaard** [comment]
@sebwrede 
Now it should work all the time with any tolerance. If you want to try again!

Otherwise if it LGT(you) then i will merge.

**Review by sebwrede — COMMENTED** [review]
Final Review, it generally looks good to me. I have a few comments (see them inline), but it can be merged after that. 
One additional comment: please change "java" to "Java" and "python" to "Python" (unless it is a command).

**Pull request closed without merging**